tracing-appender/src/worker.rs

```diff
@@ -0,0 +1,96 @@
+use crossbeam_channel::{Receiver, RecvError, TryRecvError};
+use std::fmt::Debug;
+use std::io::Write;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::{io, thread};
+
+pub(crate) struct Worker<T: Write + Send + Sync + 'static> {
+    writer: T,
+    receiver: Receiver<Vec<u8>>,
+    shutdown_signal: Arc<AtomicBool>,
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub(crate) enum WorkerState {
+    Empty,
+    Disconnected,
+    Continue,
+}
+
+impl<T: Write + Send + Sync + 'static> Worker<T> {
+    pub(crate) fn new(
+        receiver: Receiver<Vec<u8>>,
+        writer: T,
+        shutdown_signal: Arc<AtomicBool>,
+    ) -> Worker<T> {
+        Self {
+            writer,
+            receiver,
+            shutdown_signal,
+        }
+    }
+
+    fn handle_recv(&mut self, result: &Result<Vec<u8>, RecvError>) -> io::Result<WorkerState> {
+        match result {
+            Ok(msg) => {
+                self.writer.write(&msg)?;
+                Ok(WorkerState::Continue)
+            }
+            Err(_) => Ok(WorkerState::Disconnected),
+        }
+    }
+
+    fn handle_try_recv(
+        &mut self,
+        result: &Result<Vec<u8>, TryRecvError>,
+    ) -> io::Result<WorkerState> {
+        match result {
+            Ok(msg) => match self.writer.write(&msg) {
+                Ok(_) => Ok(WorkerState::Continue),
+                Err(e) => Err(e),
+            },
+            Err(e) => match e {
+                TryRecvError::Empty => Ok(WorkerState::Empty),
+                TryRecvError::Disconnected => Ok(WorkerState::Disconnected),
+            },
```

Comment:
nit: i don't think the second nested match is necessary. consider:
```suggestion
            Err(TryRecvError::Empty) => Ok(WorkerState::Empty),
            Err(TryRecvError::Disconnected) => Ok(WorkerState::Disconnected),
```